Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,308 @@
+#### QueueProcessMessage
+- Invokes the `DispatchAsync` method of the channel dispatcher by passing `QueueMessageContext` object and ending the middleware pipeline.
+- As a part of `DispatchAsync`, the service operation is called, processing of messages is done, eventually `QueueMessageContext` (inherits `RequestContext`) `ReplyAsync` method is called.
+- The `ReplyAsync` mechanism is used to inform the concrete transport that a message processing has finished with a status of success or failure by invoking an Action `DispatchResultHandler`. 
+- By calling the `DispatchResultHandler` each queue providers which support timeouts or requires notification to prevent reprocessing a message can take appropriate actions. 
+- The reply message sent via `DispatchResultHandler` could be a fault message, which will signal a failure to process the message, and can be used to provide an error message back to the queue provider.
+
+# Workflow Diagram
+![CoreWCF Generic Queue](/Documentation/DesignDocs/corewcf_queue.png?raw=true)
+
+
+ - During the server registrations, the QueuePollingService initializes the service dispatcher and gets the concrete transport from each queue layer transport.
+ - When the ASP.NET Core WebHost starts, it calls IHostedService. StartAsync on the QueuePollingService, creates a list of tasks (StartFetchingMessage) for each transport and runs them in a loop.
+ - Each task again creates a list of subtasks based on the QueueTransport (concurrent property) to retrieve message context(QueueMessageContext) (until server stops)
+ - The fetched QueueMessageContext is passed to the middleware. The first middleware layer reads the message bytes from a PipeReader and decodes them to a Message object.
+ - The Message object is put into the QueueMessageContext object and passed to the next layer, QueueProcessMessage.
+ - The QueueProcessMessage reads the message and invokes DispatchAsync on the IServiceChannelDispatcher to call the service based on the binding and message header information.
+ - Once the service is processed, QueueMessageContext sends the result to concrete transport to notify that the message is being processed. It will pass status (failed/success) and a full message , so that concrete queue transport can take appropriate actions.
+
+# Notifying Client
+- Every internal transport when constructing a `QueueMessageContext` via `ReceiveQueueMessageContextAsync` method, they can pass a delegate in the form of `DispatchResultHandler` .
+- Once a message is processed and service is invoked, the control comes to `ReplyAsync` method of `QueueMessageContext`, where if `DispatchResultHandler` is defined, it's invoked by passing a) status of service invocation and b)full `QueueMessageContext`
+- Internal transport can act based on that to notify client or put into dead letter queue or trigger any custom fuctions.
```

Comment:
...method of `QueueMessageContext`. If `DispatchResultHandler` has been set, it's invoked...  
Defined would mean whether the property or field exists on the class, not whether it holds a value. Also don't use a) and b) to list passed parameters, just list them with commas.